sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/db/migration/V8520__Attachments.java

```diff
@@ -0,0 +1,173 @@
+      this.joinTable = joinTable;
+    }
+
+    public String getDirName() {
+      return dirName;
+    }
+
+    public String getTable() {
+      return table;
+    }
+
+    public String getPrimaryKey() {
+      return primaryKey;
+    }
+
+    public String getJoinTable() {
+      return joinTable;
+    }
+  }
+
+  private Connection connection;
+  private String filesDir;
+  private Path basePath;
+
+  @Override
+  public void migrate(Connection connection) throws Exception {
+    log.info("Scanning existing attachments to save in database...");
+
+    this.connection = connection;
+    filesDir = getFilesDir();
+    if (LimsUtils.isStringEmptyOrNull(filesDir)) {
+      throw new IllegalStateException("MISO_FILES_DIR not set");
+    }
+    if (!filesDir.endsWith(File.separator)) {
+      filesDir += File.separator;
+    }
+    basePath = Paths.get(filesDir);
+
+    log.info(String.format("scanning directory %s", filesDir));
+    try (PreparedStatement insertStatement = connection.prepareStatement(
+        "INSERT INTO Attachment(filename, path, creator, created)"
+            + " VALUES (?, ?, (SELECT userId FROM User WHERE loginName = 'admin'), NOW());",
+        Statement.RETURN_GENERATED_KEYS)) {
+      for (EntityType type : EntityType.values()) {
+        scanTypeDir(type, insertStatement);
+      }
+    }
+
+    dropTempTable();
+  }
+
+  private void scanTypeDir(EntityType type, PreparedStatement insertStatement) throws SQLException {
+    File dir = new File(filesDir + type.getDirName());
+    if (dir.exists()) {
+      log.info(String.format("scanning directory %s", dir.getAbsolutePath()));
+      try (PreparedStatement getStatement = connection
+          .prepareStatement(String.format("SELECT * FROM %s WHERE %s = ?", type.getTable(), type.getPrimaryKey()));
+          PreparedStatement joinStatement = connection.prepareStatement(
+              String.format("INSERT INTO %s(%s, attachmentId) VALUES (?, ?);", type.getJoinTable(), type.getPrimaryKey()))) {
+        for (File typeFile : dir.listFiles()) {
+          processObjectDir(type, typeFile, getStatement, insertStatement, joinStatement);
+        }
+      }
+    }
+  }
+
+  private void processObjectDir(EntityType type, File dir, PreparedStatement getStatement, PreparedStatement insertStatement,
+      PreparedStatement joinStatement) throws SQLException {
+    log.info("scanning directory " + dir.getAbsolutePath());
+    if (!dir.isDirectory()) {
+      log.warning(String.format("Unexpected non-directory found: %s", dir.getAbsolutePath()));
+    } else {
+      long entityId = 0L;
+      try {
+        entityId = Long.parseLong(dir.getName());
+      } catch (NumberFormatException e) {
+        // ignore error - failure handled below
+      }
+      if (entityId == 0L) {
+        log.warning(String.format("Unexpected directory name: %s", dir.getAbsolutePath()));
+      } else {
+        getStatement.setLong(1, entityId);
+        if (!getStatement.executeQuery().next()) {
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "ResultSet" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2095)